searx/network/client.py

```diff
@@ -155,56 +192,377 @@ def get_transport(verify, http2, local_address, proxy_url, limit, retries):
+            max_redirects = self.default.max_redirects
+        return _HTTPMultiClientConf(verify, max_redirects), kwargs
 
-    transport = get_transport(verify, enable_http2, local_address, None, limit, retries)
 
-    event_hooks = None
-    if hook_log_response:
-        event_hooks = {'response': [hook_log_response]}
+class HTTPClient(BaseHTTPClient):
+    """Inherit from BaseHTTPClient, raise an exception according to the retry_on_http_error parameter"""
 
-    return httpx.AsyncClient(
-        transport=transport,
-        mounts=mounts,
-        max_redirects=max_redirects,
-        event_hooks=event_hooks,
-    )
+    def __init__(self, retry_on_http_error=None, **kwargs):
+        super().__init__(**kwargs)
+        self.retry_on_http_error = retry_on_http_error
+        self._check_configuration()
+
+    def _check_configuration(self):
+        # make sure we can create at least an OneHTTPClient without exception
+        self._get_client_and_update_kwargs({})
 
+    def send(self, stream, method, url, timeout=None, **kwargs):
+        try:
+            do_raise_for_httperror = self._extract_do_raise_for_httperror(kwargs)
+            response = super().send(stream, method, url, timeout=timeout, **kwargs)
+            if do_raise_for_httperror:
+                raise_for_httperror(response)
+            if self._is_error_but_retry(response):
+                raise SoftRetryHTTPException(response)
+            return response
+        except (httpx.RequestError, httpx.HTTPStatusError) as e:
+            raise e
+
+    def _is_error_but_retry(self, response):
+        # pylint: disable=too-many-boolean-expressions
+        return (
+            (self.retry_on_http_error is True and 400 <= response.status_code <= 599)
+            or (isinstance(self.retry_on_http_error, list) and response.status_code in self.retry_on_http_error)
+            or (isinstance(self.retry_on_http_error, int) and response.status_code == self.retry_on_http_error)
+        )
+
+    @staticmethod
+    def _extract_do_raise_for_httperror(kwargs):
+        do_raise_for_httperror = True
+        if 'raise_for_httperror' in kwargs:
+            do_raise_for_httperror = kwargs['raise_for_httperror']
+            del kwargs['raise_for_httperror']
+        return do_raise_for_httperror
+
+    def __repr__(self):
+        keys_values = " ".join([f"{k}={v!r}" for k, v in self.default_kwargs.items()])
+        return f"<{self.__class__.__name__} retry_on_http_error={self.retry_on_http_error!r} {keys_values}>"
+
+
+class TorHTTPClient(HTTPClient):
+    """Extend HTTPClientSoftError client. To use with Tor configuration.
+
+    The class checks if the client is really connected through Tor.
+    """
+
+    _TOR_CHECK_RESULT = {}
+
+    def __init__(self, proxies=None, local_addresses=None, **kwargs):
+        self.proxies = proxies
+        self.local_addresses = local_addresses
+        super().__init__(proxies=proxies, local_addresses=local_addresses, **kwargs)
```

Comment:
@dalf I'm wrong, or can we drop this TOR constructor and simply use the constructor of the parent class? If so, I remove it in my review ..

Reply:
Actually yes:
* drop `__init__`
* replace https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/client.py#L516 by
`if not self._is_connected_through_tor(self.default_kwargs["proxies"], self.default_kwargs.get("local_addresses")):`

it should work